### ISSUE TITLE
SkipLocalsInit condition applies to NetCoreAppCurrent only

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#54964]
See https://github.com/dotnet/runtime/blob/a18b2cbd06c3b53c8d42d38b3c859079003576b6/src/libraries/Directory.Build.targets#L273

The SkipLocalsInit property is conditioned on `NetCoreAppCurrent` inbox assets only. Some packages contain both inbox and OOB assets. These optimization will be lost for net6.0 when `NetCoreAppCurrent` means net7.0 and new packages will be released.

One option would be to support SkipLocalsInit on all .NETCoreApp tfms. Another option would be to not apply SkipLocalsInit to assets that ship in packages as well but which would cause a perf regression.

cc @stephentoub @ericstj @jkotas @eerhardt

**Comment by stephentoub** [comment]
> The SkipLocalsInit property is conditioned on NetCoreAppCurrent inbox assets only

This correctly reflects the original intent.  We audited the libs in the shared framework; we didn't do so for OOBs.

>  Some packages contain both inbox and OOB assets.

Can you elaborate on that?  What does it mean for a package to contain an inbox asset?

> One option would be to support SkipLocalsInit on all .NETCoreApp tfms.

We would need to go through all of our assets that this would apply to, in particular ones with stackallocs, P/Invokes, and in general unsafe code, and make sure everything is sufficiently protected.

> Another option would be to not apply SkipLocalsInit to assets that ship in packages as well but which would cause a perf regression.

This is where we should start, and then opt things in as we find opportunities / meaningful improvements.

**Comment by joperezr** [comment]
>Can you elaborate on that? What does it mean for a package to contain an inbox asset?

I believe that what @ViktorHofer meant here, is that there are some inbox assets that also get shipped via its library's OOB package, for example System.Text.Json. It's net6.0 asset ships inbox, but also gets shipped in the System.Text.Json package since we want to be able to replace it if we want to via the package, and I can see how @ViktorHofer concern is valid in case we were shipping a slightly different asset inbox than the OOB one. That said, I believe that is not the case  here, as if I understand correctly then the OOB asset should also use SkipLocalsInit as the "IsNETCoreAppSrc" condition will still evaluate to true.

@ViktorHofer given the answers from @stephentoub above, are there still concerns with the way this feature works today? If not, let's move this issue out of 6.0.0 in order to track in the future the work to try to enable the feature for assets targeting older frameworks.

**Comment by ViktorHofer** [comment]
> Can you elaborate on that? What does it mean for a package to contain an inbox asset?

6.0 packages produced in dotnet/runtime contain net6.0 assets even if those ship as part of a shared framework. One example of that is System.Text.Json. During a build, conflict resolution which part of the sdk trims out assets which are satisfied by the shared framework. Conflict resolution takes the PlatformManifest which lists all assemblies including their assembly and file versions as an inputs and compares those against references which are included i.e. via a package.

In the event of servicing, a change in System.Text.Json is delivered via the shared framework and/or via the updated System.Text.Json package. Conflict resolution will make sure that the best compatible asset is chosen.

> This correctly reflects the original intent. We audited the libs in the shared framework; we didn't do so for OOBs.

Consider the 7.0 package of System.Text.Json which will include both the net6.0 asset as well as the net7.0 asset (according to our package support policy). With the current condition: 

`<SkipLocalsInit Condition="'$(SkipLocalsInit)' == '' and '$(MSBuildProjectExtension)' == '.csproj' and '$(IsNETCoreAppSrc)' == 'true' and ($([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', '$(NetCoreAppCurrent)')))">true</SkipLocalsInit>` 

`SkipLocalsInit` is only applied to `$(NetCoreAppCurrent)` which maps to net7.0 when producing a 7.0 package. As the net6.0 asset is live built, it will loose the `SkipLocalsInit` optimization which it received when `$(NetCoreAppCurrent)` mapped to net6.0. 
My assumption is that we don't want to regress the previous shipped asset by changing the `SkipLocalsInit` condition to:

`<SkipLocalsInit Condition="'$(SkipLocalsInit)' == '' and '$(MSBuildProjectExtension)' == '.csproj' and '$(IsNETCoreAppSrc)' == 'true'">true</SkipLocalsInit>`

That way the optimization will only apply to assemblies which are part of the shared framework either in the current release or in the previous release.

**Comment by stephentoub** [comment]
> My assumption is that we don't want to regress the previous shipped asset

Yes, if that's what's going to happen, we should fix that.

**Comment by stephentoub** [comment]
```C#
    [Benchmark]
    public int GetValue() => GetValue(42);

    [MethodImpl(MethodImplOptions.NoInlining)]
    private int GetValue(int? n)
    {
        if (n.HasValue) return n.Value;
        return -1;
    }
```

.NET 5:
```assembly
; Program.GetValue(System.Nullable`1<Int32>)
       sub       rsp,28
       mov       [rsp+38],rdx
       movzx     eax,byte ptr [rsp+38]
       test      eax,eax
       je        short M01_L00
       test      eax,eax
       je        short M01_L01
       mov       eax,[rsp+3C]
       add       rsp,28
       ret
M01_L00:
       mov       eax,0FFFFFFFF
       add       rsp,28
       ret
M01_L01:
       call      System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_NoValue()
       int       3
; Total bytes of code 47
```

.NET 6:
```assembly
; Program.GetValue(System.Nullable`1<Int32>)
       mov       [rsp+10],rdx
       cmp       byte ptr [rsp+10],0
       je        short M01_L00
       mov       eax,[rsp+14]
       ret
M01_L00:
       mov       eax,0FFFFFFFF
       ret
; Total bytes of code 23
```

**Comment by ViktorHofer** [comment]
@stephentoub why did you close the issue? The issue is present in the main branch and needs to be fixed to avoid 7.0 packages regress net6.0 assets.

**Comment by stephentoub** [comment]
Shoot.  I posted my comment on and closed the wrong issue.  Too many tabs open :)

**Comment by ViktorHofer** [comment]
no worries

**Comment by ViktorHofer** [comment]
Adding the issue back to 6.0 as I believe this needs to be fixed in both main and release/6.0-rc1.

**Comment by ViktorHofer** [comment]
Retargeting to 7.0 as the remaining work is main specific. After https://github.com/dotnet/runtime/issues/57844 is in, this will be a matter of updating the condition and removing the last `and ...` part.
